keps/sig-api-machinery/4192-svm-in-tree/README.md

```diff
@@ -0,0 +1,483 @@
+
+### Goals
+
+- Make it easy for an admin to trigger a storage migration without having to manually run `kubectl get | kubectl replace`
+- Enable infrastructure providers to use internal details of their environment to trigger storage migration (ex: triggering storage migration of Kubernetes secrets when the KMS v2 key ID has changed)
+- Make it easy for existing users of SVM to migrate to the in-tree controller
+
+### Non-Goals
+
+- Any automatic or direct integration with KMS v2
+- Any modification regarding `StorageVersion` API for HA API servers
+- Adding logic that relies on the hashed storage versions exposed via the discovery API
+
+### UNCLEAR Goals and/or Non-Goals
+
+- Automatic storage version migration for CRDs
+- Make it easy for Kubernetes developers to drop old API schemas by guaranteeing that storage migration is run automatically on SV hash changes (should this also be on a timer or API server identity?)
+- Automated Storage Version Migration via the hash exposed by the `StorageVersion` API
+
+## Proposal
+
+- Move the existing SVM controller logic in-tree into KCM
+- Move the existing SVM REST APIs in-tree (possibly under a new API group to avoid conflicts with the old API being run concurrently)
+- **[UNCLEAR]** Create a new controller that watches the storage version API and automatically triggers migrations based on changes (to the SV API)
+
+### User Stories (Optional)
+
+#### Story 1 **[UNCLEAR]**
+As an end user of Kubernetes, we get automatic storage version migration whenever the storage version changes due to an api server upgrade/downgrade.
+
+#### Story 2
+As an end user using encryption at rest, whenever the key change is detected we can run the storage migration to use the new key for encryption. 
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+## Design Details
+
+### APIs to move
+#### We will move following [APIs](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/60dee538334c2366994c2323c0db5db8ab4d2838/pkg/apis/migration/v1alpha1/types.go) in-tree:
+- `v1alpha1` of `storageversionmigrations.migration.k8s.io`  
+
+    ```go
+    // StorageVersionMigration represents a migration of stored data to the latest
+    // storage version.
+    type StorageVersionMigration struct {
+        metav1.TypeMeta `json:",inline"`
+        // +optional
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        // Specification of the migration.
+        // +optional
+        Spec StorageVersionMigrationSpec `json:"spec,omitempty"`
+        // Status of the migration.
+        // +optional
+        Status StorageVersionMigrationStatus `json:"status,omitempty"`
+    }
+
+    // Spec of the storage version migration.
+    type StorageVersionMigrationSpec struct {
+        // The resource that is being migrated. The migrator sends requests to
+        // the endpoint serving the resource.
+        // Immutable.
+        Resource GroupVersionResource `json:"resource"`
```

Comment:
When reading this it sounded like this is the GVR being migrated "from". Is that correct? Should we clarify further?

Reply:
I updated the comment and the field name as well. PTAL. I'm open to other suggestions as well. 😄